### ISSUE TITLE
List tryRemove function

### ISSUE THREAD

**Issue by aaron-jack-manning** [fsharp/fslang-suggestions#1093]
# List tryRemove function

My proposal is for a function in the List module (and a similar function for the other collections) to remove the first instance of a specified element in a list, which falls back gracefully when the element isn't found by returning the original list.

## Sample code

```
let rec tryRemove element list =
    match list with
    | [] -> []
    | h :: t -> if h = element then t else h :: tryRemove element t
```

Currently, I believe the neatest implementation of this when writing F# code is to write a function like the above, as opposed to using standard library functions such as `tryFindIndex` and `removeAt`, hence the suggestion for it to exist as its own function.

## Considerations

Some consideration should be given to:
- the name of such a function, since many functions with the `try` prefix return option type, which the above does not, but it doesn't do so without throwing an exception either.
- if alternative, similar functions should exist which throw exceptions or return option type to handle cases where the specified element is not found.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [X] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
* [X] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by Frassle** [comment]
Note that both Map and Set have functions called just `remove` that behave like this. If the element is not found it just returns the input collection. I'd suggest for consistency this would also be called `remove`.

**Comment by aaron-jack-manning** [comment]
> Note that both Map and Set have functions called just `remove` that behave like this. If the element is not found it just returns the input collection. I'd suggest for consistency this would also be called `remove`.

I didn't realise that. That makes the issue of naming easy to handle, although some care should be made in documentation since Set guarantees uniqueness of elements whereas in this case only the first element will be removed, not all of them, since to remove all you could just use `List.where`.

**Comment by Happypig375** [comment]
But why specifically the first element? What are the use cases?

**Comment by aaron-jack-manning** [comment]
> But why specifically the first element? What are the use cases?

I've needed a function like this many times. [Here](https://gist.github.com/aaron-jack-manning/652598c5c3e2732219bd2cd377a07389) is one such example. In that case, the list represents a sequence of items that needs to be exhausted, and so having used one, it must be removed. This is because the order in the list isn't really important. However, ultimately the reason to remove the first is because the goal is to remove an element by its value rather than its index (which already exists), and that seems like the most sensible behaviour for such a function (noting again that removing all occurrences can easily be done using `List.where`).

**Comment by Happypig375** [comment]
Ah, using a list like a stack.

**Comment by Happypig375** [comment]
What about `removeFirst` instead of `remove`?

**Comment by aaron-jack-manning** [comment]
Not quite like a stack, because in a stack you wouldn't look through to remove a certain element. Only like a stack if the list is in the correct order. The only issue with `removeFirst` is to me that reads like removing the first element of the list, rather than removing the specified element, which is what I'm proposing.

**Comment by Tarmil** [comment]
I think that just calling it `remove` is fine because there is precedent: `find`, `tryFind`, `pick` and `tryPick` return the first matching occurrence.

**Comment by Happypig375** [comment]
@Tarmil The signatures of those functions return one element or option of one element. `remove` has no such indication.

**Comment by aaron-jack-manning** [comment]
Yes, agreed. The naming inconsistency in an issue here. Maybe `removeOne` or something?
